src/main/java/cnlabs/CustomerWebsite/Controllers/CustomerController.java

```diff
@@ -19,46 +21,44 @@ public class CustomerController {
 
     @GetMapping("/")
     public String viewHomePage(Model model) {
-        // Here you call the service to retrieve all the customers
         final List<Customer> customerList = customerService.getAllCustomers();
-        // Once the customers are retrieved, you can store them in model and return it to the view
         model.addAttribute("customerList", customerList);
         return "index";
     }
 
     @GetMapping("/new")
     public String showNewCustomerPage(Model model) {
-        // Here a new (empty) Customer is created and then sent to the view
         Customer customer = new Customer();
         model.addAttribute("customer", customer);
         return "new-customer";
     }
 
     @PostMapping("/save")
-    // As the Model is received back from the view, @ModelAttribute
-    // creates a Customer based on the object you collected from
-    // the HTML page above
-    public String saveCustomer(@ModelAttribute("customer") Customer customer) {
+    public String saveCustomer(@Valid @ModelAttribute("customer") Customer customer, BindingResult result, Model model) {
+        if (result.hasErrors()) {
+            return "new-customer";
+        }
         customerService.saveCustomer(customer);
         return "redirect:/";
     }
 
     @GetMapping("/edit/{id}")
-    // The path variable "id" is used to pull a customer from the database
     public ModelAndView showEditCustomerPage(@PathVariable(name = "id") Long id) {
-        // Since the previous methods use Model, this one uses ModelAndView
-        // to get some experience using both. Model is more common these days,
-        // but ModelAndView accomplishes the same thing and can be useful in
-        // certain circumstances. The view name is passed to the constructor.
         ModelAndView mav = new ModelAndView("edit-customer");
         Customer customer = customerService.getCustomer(id);
+        if (customer == null) {  // If some error happens while trying to get the customer, return home.
+            return new ModelAndView("redirect:/");
+        }
         mav.addObject("customer", customer);
         return mav;
     }
 
     @PostMapping("/update/{id}")
-    public String updateCustomer(@PathVariable(name = "id") Long id, @ModelAttribute("customer") Customer customer, Model model) {
-        if (!id.equals(customer.getId())) {
+    public String updateCustomer(@PathVariable(name = "id") Long id, @ModelAttribute("customer") Customer customer, BindingResult result, Model model) {
+        if (result.hasErrors()) {
+            return "/edit/{id}";
+        }
+        if (!id.equals(customer.getId())) {  // I'm not sure how this would ever get triggered. The user would have to edit the post request themselves.
```

Comment:
Postman can do it easilly, `curl` will do the trick if you know it. Even intellij can do it, i'll show you next time.

Reply:
Right! Because public endpoints can be accessed from anywhere, not just within the application?